carsus/io/output/base.py

```diff
@@ -0,0 +1,546 @@
+            lambda x: x.to('angstrom'))
+        lines['wavelength'] = lines['wavelength'].apply(lambda x: x.value)
+
+        air_mask = lines['medium'] == MEDIUM_AIR
+        lines.loc[air_mask, 'wavelength'] = convert_wavelength_air2vacuum(
+            lines.loc[air_mask, 'wavelength'])
+        lines.drop(columns=['medium'], inplace=True)
+        lines = lines[['lower_level_id', 'upper_level_id',
+                       'wavelength', 'gf', 'loggf']]
+
+        return lines
+
+    def _create_levels_lines(self, lines_loggf_threshold=-3,
+                             levels_metastable_loggf_threshold=-3):
+        """ Returns almost the same output than
+        `AtomData.create_levels_lines` method """
+        levels_all = self.levels_all
+        lines_all = self.lines_all
+        ionization_energies = self.ionization_energies.reset_index()
+        ionization_energies['ion_number'] -= 1
+
+        # Culling autoionization levels
+        levels_w_ionization_energies = pd.merge(levels_all,
+                                                ionization_energies,
+                                                how='left',
+                                                on=["atomic_number",
+                                                    "ion_number"])
+        mask = levels_w_ionization_energies["energy"] < \
+            levels_w_ionization_energies["ionization_energy"]
+        levels = levels_w_ionization_energies[mask].copy()
+        levels = levels.set_index('level_id').sort_values(
+            by=['atomic_number', 'ion_number'])
+        levels = levels.drop(columns='ionization_energy')
+
+        # Clean lines
+        lines = lines_all.join(pd.DataFrame(index=levels.index),
+                               on="lower_level_id", how="inner").\
+            join(pd.DataFrame(index=levels.index),
+                 on="upper_level_id", how="inner")
+
+        # Culling lines with low gf values
+        lines = lines.loc[lines["loggf"] > lines_loggf_threshold]
+
+        # Do not clean levels that don't exist in lines
+
+        # Create the metastable flags for levels
+        levels["metastable"] = \
+            self._create_metastable_flags(levels, lines_all,
+                                          levels_metastable_loggf_threshold)
+
+        # Create levels numbers
+        levels = levels.sort_values(
+            ["atomic_number", "ion_number", "energy", "g"])
+        levels["level_number"] = levels.groupby(['atomic_number',
+                                                 'ion_number'])['energy']. \
+            transform(lambda x: np.arange(len(x))).values
+        levels["level_number"] = levels["level_number"].astype(np.int)
+
+        levels = levels[['atomic_number', 'energy', 'g', 'ion_number',
+                         'level_number', 'metastable']]
+
+        # Join atomic_number, ion_number, level_number_lower,
+        # level_number_upper on lines
+        lower_levels = levels.rename(
```

Comment:
This is quite difficult to read. Maybe moving the dictionary with the columns out of the expression might help.